### ISSUE TITLE
Tests crash on Python 3.5

### ISSUE THREAD

**Issue by jaraco** [yougov/pmxbot#42]
When collecting tests on Python 3.5, something happens that causes runaway memory consumption and crashes the tests. See [this build for example](https://travis-ci.org/jaraco/pmxbot/builds/94400144).

----------------------------------------
- Bitbucket: https://bitbucket.org/yougov/pmxbot/issue/42
- Originally reported by: Jason R. Coombs
- Originally created at: 2015-12-02T15:12:48.616

**Comment by jaraco** [comment]
I managed to kill the process after it had started to runaway, and here's the traceback I got:

```
Traceback (most recent call last):
  File "/Users/jaraco/.envs/pmxbot/lib/python3.5/site-packages/_pytest/main.py", line 90, in wrap_session
    session.exitstatus = doit(config, session) or 0
  File "/Users/jaraco/.envs/pmxbot/lib/python3.5/site-packages/_pytest/main.py", line 120, in _main
    config.hook.pytest_collection(session=session)
  File "/Users/jaraco/.envs/pmxbot/lib/python3.5/site-packages/_pytest/vendored_packages/pluggy.py", line 724, in __call__
    return self._hookexec(self, self._nonwrappers + self._wrappers, kwargs)
  File "/Users/jaraco/.envs/pmxbot/lib/python3.5/site-packages/_pytest/vendored_packages/pluggy.py", line 338, in _hookexec
    return self._inner_hookexec(hook, methods, kwargs)
  File "/Users/jaraco/.envs/pmxbot/lib/python3.5/site-packages/_pytest/vendored_packages/pluggy.py", line 333, in <lambda>
    _MultiCall(methods, kwargs, hook.spec_opts).execute()
  File "/Users/jaraco/.envs/pmxbot/lib/python3.5/site-packages/_pytest/vendored_packages/pluggy.py", line 596, in execute
    res = hook_impl.function(*args)
  File "/Users/jaraco/.envs/pmxbot/lib/python3.5/site-packages/_pytest/main.py", line 129, in pytest_collection
    return session.perform_collect()
  File "/Users/jaraco/.envs/pmxbot/lib/python3.5/site-packages/_pytest/main.py", line 561, in perform_collect
    items = self._perform_collect(args, genitems)
  File "/Users/jaraco/.envs/pmxbot/lib/python3.5/site-packages/_pytest/main.py", line 597, in _perform_collect
    self.items.extend(self.genitems(node))
  File "/Users/jaraco/.envs/pmxbot/lib/python3.5/site-packages/_pytest/main.py", line 734, in genitems
    rep = collect_one_node(node)
  File "/Users/jaraco/.envs/pmxbot/lib/python3.5/site-packages/_pytest/runner.py", line 411, in collect_one_node
    rep = ihook.pytest_make_collect_report(collector=collector)
  File "/Users/jaraco/.envs/pmxbot/lib/python3.5/site-packages/_pytest/vendored_packages/pluggy.py", line 724, in __call__
    return self._hookexec(self, self._nonwrappers + self._wrappers, kwargs)
  File "/Users/jaraco/.envs/pmxbot/lib/python3.5/site-packages/_pytest/vendored_packages/pluggy.py", line 338, in _hookexec
    return self._inner_hookexec(hook, methods, kwargs)
  File "/Users/jaraco/.envs/pmxbot/lib/python3.5/site-packages/_pytest/vendored_packages/pluggy.py", line 333, in <lambda>
    _MultiCall(methods, kwargs, hook.spec_opts).execute()
  File "/Users/jaraco/.envs/pmxbot/lib/python3.5/site-packages/_pytest/vendored_packages/pluggy.py", line 595, in execute
    return _wrapped_call(hook_impl.function(*args), self.execute)
  File "/Users/jaraco/.envs/pmxbot/lib/python3.5/site-packages/_pytest/vendored_packages/pluggy.py", line 249, in _wrapped_call
    wrap_controller.send(call_outcome)
  File "/Users/jaraco/.envs/pmxbot/lib/python3.5/site-packages/_pytest/capture.py", line 112, in pytest_make_collect_report
    rep = outcome.get_result()
  File "/Users/jaraco/.envs/pmxbot/lib/python3.5/site-packages/_pytest/vendored_packages/pluggy.py", line 278, in get_result
    raise ex[1].with_traceback(ex[2])
  File "/Users/jaraco/.envs/pmxbot/lib/python3.5/site-packages/_pytest/vendored_packages/pluggy.py", line 264, in __init__
    self.result = func()
  File "/Users/jaraco/.envs/pmxbot/lib/python3.5/site-packages/_pytest/vendored_packages/pluggy.py", line 596, in execute
    res = hook_impl.function(*args)
  File "/Users/jaraco/.envs/pmxbot/lib/python3.5/site-packages/_pytest/runner.py", line 283, in pytest_make_collect_report
    call = CallInfo(collector._memocollect, "memocollect")
  File "/Users/jaraco/.envs/pmxbot/lib/python3.5/site-packages/_pytest/runner.py", line 149, in __init__
    self.result = func()
  File "/Users/jaraco/.envs/pmxbot/lib/python3.5/site-packages/_pytest/main.py", line 431, in _memocollect
    return self._memoizedcall('_collected', lambda: list(self.collect()))
  File "/Users/jaraco/.envs/pmxbot/lib/python3.5/site-packages/_pytest/main.py", line 311, in _memoizedcall
    res = function()
  File "/Users/jaraco/.envs/pmxbot/lib/python3.5/site-packages/_pytest/main.py", line 431, in <lambda>
    return self._memoizedcall('_collected', lambda: list(self.collect()))
  File "/Users/jaraco/.envs/pmxbot/lib/python3.5/site-packages/_pytest/doctest.py", line 186, in collect
    for test in finder.find(module, module.__name__):
  File "/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/doctest.py", line 923, in find
    self._find(tests, obj, name, module, source_lines, globs, {})
  File "/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/doctest.py", line 982, in _find
    if ((inspect.isroutine(inspect.unwrap(val))
  File "/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/inspect.py", line 471, in unwrap
    while _is_wrapper(func):
  File "/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/inspect.py", line 465, in _is_wrapper
    return hasattr(f, '__wrapped__')
  File "/Users/jaraco/.envs/pmxbot/lib/python3.5/site-packages/mock/mock.py", line 2204, in __getattr__
    return _Call(name=name, parent=self, from_kall=False)
  File "/Users/jaraco/.envs/pmxbot/lib/python3.5/site-packages/mock/mock.py", line 2098, in __new__
    def __new__(cls, value=(), name=None, parent=None, two=False,
KeyboardInterrupt

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/jaraco/.envs/pmxbot/bin/py.test", line 11, in <module>
    sys.exit(main())
  File "/Users/jaraco/.envs/pmxbot/lib/python3.5/site-packages/_pytest/config.py", line 48, in main
    return config.hook.pytest_cmdline_main(config=config)
  File "/Users/jaraco/.envs/pmxbot/lib/python3.5/site-packages/_pytest/vendored_packages/pluggy.py", line 724, in __call__
    return self._hookexec(self, self._nonwrappers + self._wrappers, kwargs)
  File "/Users/jaraco/.envs/pmxbot/lib/python3.5/site-packages/_pytest/vendored_packages/pluggy.py", line 338, in _hookexec
    return self._inner_hookexec(hook, methods, kwargs)
  File "/Users/jaraco/.envs/pmxbot/lib/python3.5/site-packages/_pytest/vendored_packages/pluggy.py", line 333, in <lambda>
    _MultiCall(methods, kwargs, hook.spec_opts).execute()
  File "/Users/jaraco/.envs/pmxbot/lib/python3.5/site-packages/_pytest/vendored_packages/pluggy.py", line 596, in execute
    res = hook_impl.function(*args)
  File "/Users/jaraco/.envs/pmxbot/lib/python3.5/site-packages/_pytest/main.py", line 115, in pytest_cmdline_main
    return wrap_session(config, _main)
  File "/Users/jaraco/.envs/pmxbot/lib/python3.5/site-packages/_pytest/main.py", line 95, in wrap_session
    config.hook.pytest_keyboard_interrupt(excinfo=excinfo)
  File "/Users/jaraco/.envs/pmxbot/lib/python3.5/site-packages/_pytest/vendored_packages/pluggy.py", line 724, in __call__
    return self._hookexec(self, self._nonwrappers + self._wrappers, kwargs)
  File "/Users/jaraco/.envs/pmxbot/lib/python3.5/site-packages/_pytest/vendored_packages/pluggy.py", line 338, in _hookexec
    return self._inner_hookexec(hook, methods, kwargs)
  File "/Users/jaraco/.envs/pmxbot/lib/python3.5/site-packages/_pytest/vendored_packages/pluggy.py", line 333, in <lambda>
    _MultiCall(methods, kwargs, hook.spec_opts).execute()
  File "/Users/jaraco/.envs/pmxbot/lib/python3.5/site-packages/_pytest/vendored_packages/pluggy.py", line 596, in execute
    res = hook_impl.function(*args)
  File "/Users/jaraco/.envs/pmxbot/lib/python3.5/site-packages/_pytest/terminal.py", line 378, in pytest_keyboard_interrupt
    self._keyboardinterrupt_memo = excinfo.getrepr(funcargs=True)
  File "/Users/jaraco/.envs/pmxbot/lib/python3.5/site-packages/py/_code/code.py", line 412, in getrepr
    return fmt.repr_excinfo(self)
  File "/Users/jaraco/.envs/pmxbot/lib/python3.5/site-packages/py/_code/code.py", line 590, in repr_excinfo
    reprtraceback = self.repr_traceback(excinfo)
  File "/Users/jaraco/.envs/pmxbot/lib/python3.5/site-packages/py/_code/code.py", line 582, in repr_traceback
    reprentry = self.repr_traceback_entry(entry, einfo)
  File "/Users/jaraco/.envs/pmxbot/lib/python3.5/site-packages/py/_code/code.py", line 528, in repr_traceback_entry
    source = self._getentrysource(entry)
  File "/Users/jaraco/.envs/pmxbot/lib/python3.5/site-packages/py/_code/code.py", line 455, in _getentrysource
    source = entry.getsource(self.astcache)
  File "/Users/jaraco/.envs/pmxbot/lib/python3.5/site-packages/py/_code/code.py", line 204, in getsource
    astnode=astnode)
  File "/Users/jaraco/.envs/pmxbot/lib/python3.5/site-packages/py/_code/source.py", line 348, in getstatementrange_ast
    astnode = compile(content, "source", "exec", 1024)  # 1024 for AST
KeyboardInterrupt
```

----------------------------------------
Original comment by: Jason R. Coombs

**Comment by jaraco** [comment]
Disabling doctests bypasses the issue. Also, removing test_core.py bypasses the issue.

----------------------------------------
Original comment by: Jason R. Coombs

**Comment by jaraco** [comment]
Replacing test_core.py with simply `from mock import call` is sufficient to trigger the crash.

----------------------------------------
Original comment by: Jason R. Coombs

**Comment by jaraco** [comment]
Resolved in some commit. For some reason, bitbucket didn't pick it up.

----------------------------------------
Original comment by: Jason R. Coombs

**Comment by jaraco** [comment]
Take call out of the namespace. Workaround for https://github.com/pytest-dev/pytest/issues/1217. Fixes #42.

→ <<cset 200134da1cd7>>

----------------------------------------
Original comment by: Jason R. Coombs
